migration-engine/core/src/commands/diagnose_migration_history.rs

```diff
@@ -132,22 +134,31 @@ impl<'a> MigrationCommand for DiagnoseMigrationHistoryCommand {
             .cloned()
             .collect();
 
-        let drift = match migration_inferrer.calculate_drift(&applied_migrations).await {
-            Ok(Some(rollback)) => Some(DriftDiagnostic::DriftDetected { rollback }),
-            Err(error) => Some(DriftDiagnostic::MigrationFailedToApply {
-                error: error.to_user_facing(),
-            }),
-            _ => None,
-        };
+        let (drift, error_in_unapplied_migration) = {
+            if input.opt_in_to_shadow_database {
+                let drift = match migration_inferrer.calculate_drift(&applied_migrations).await {
+                    Ok(Some(rollback)) => Some(DriftDiagnostic::DriftDetected { rollback }),
+                    Err(error) => Some(DriftDiagnostic::MigrationFailedToApply {
+                        error: error.to_user_facing(),
+                    }),
+                    _ => None,
+                };
+
+                let error_in_unapplied_migration =
+                    if !matches!(drift, Some(DriftDiagnostic::MigrationFailedToApply { .. })) {
+                        migration_inferrer
+                            .validate_migrations(&migrations_from_filesystem)
+                            .await
+                            .err()
+                            .map(|connector_error| connector_error.to_user_facing())
+                    } else {
+                        None
+                    };
 
-        let error_in_unapplied_migration = if !matches!(drift, Some(DriftDiagnostic::MigrationFailedToApply { .. })) {
-            migration_inferrer
-                .validate_migrations(&migrations_from_filesystem)
-                .await
-                .err()
-                .map(|connector_error| connector_error.to_user_facing())
-        } else {
-            None
+                (drift, error_in_unapplied_migration)
+            } else {
+                (None, None)
+            }
```

Comment:
not a fan of that conditional... Let's see how the status command evolves, maybe we can simplify again soon.

Reply:
An alternative would be splitting this command in two, but it is too big of a change too close to release to do now.